<a href="https://colab.research.google.com/github/housekjh1/study-pandas/blob/main/Chap03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 3.1 read_csv 함수로 데이터셋 가져오기

In [1]:
import pandas as pd

read_csv 함수 : csv 파일을 읽어 DataFrame으로 반환
---
포켓몬 데이터셋

In [2]:
#pd.read_csv(filepath_or_buffer="pokemon.csv")
pd.read_csv("pokemon.csv")

Pokemon            Type
0      Bulbasaur  Grass / Poison
1        Ivysaur  Grass / Poison
2       Venusaur  Grass / Poison
3     Charmander            Fire
4     Charmeleon            Fire
..           ...             ...
804    Stakataka    Rock / Steel
805  Blacephalon    Fire / Ghost
806      Zeraora        Electric
807       Meltan           Steel
808     Melmetal           Steel

[809 rows x 2 columns]

read_csv 함수의 반환값을 DataFrame --> Series로 바꾸기
- 열 중에 하나를 Series의 인덱스로 설정하기

In [5]:
pd.read_csv("pokemon.csv", index_col="Pokemon")

Type
Pokemon                    
Bulbasaur    Grass / Poison
Ivysaur      Grass / Poison
Venusaur     Grass / Poison
Charmander             Fire
Charmeleon             Fire
...                     ...
Stakataka      Rock / Steel
Blacephalon    Fire / Ghost
Zeraora            Electric
Meltan                Steel
Melmetal              Steel

[809 rows x 1 columns]

squeeze 옵션을 True로 두어 Series로 강제 변환

In [52]:
# pd.read_csv("pokemon.csv", index_col="Pokemon", squeeze=True) # 에러발생
# pokemon = pd.read_csv("pokemon.csv", index_col="Pokemon").squeeze("columns")
pokemon = pd.read_csv("pokemon.csv", index_col="Pokemon", squeeze=True)
pokemon

<ipython-input-52-3ff15f891236>:3: FutureWarning: The squeeze argument has been deprecated and will be removed in a future version. Append .squeeze("columns") to the call to squeeze.


  pokemon = pd.read_csv("pokemon.csv", index_col="Pokemon", squeeze=True)


Pokemon
Bulbasaur      Grass / Poison
Ivysaur        Grass / Poison
Venusaur       Grass / Poison
Charmander               Fire
Charmeleon               Fire
                    ...      
Stakataka        Rock / Steel
Blacephalon      Fire / Ghost
Zeraora              Electric
Meltan                  Steel
Melmetal                Steel
Name: Type, Length: 809, dtype: object

구글의 주가 데이터

판다스는 데이터셋을 가져올 때 적합한 데이터 유형을 유추해서 가져옴.<br/>
기본적으로는 안전함을 추구함.
- yyyy-mm-dd의 경우 날짜로 명시적으로 지시하지 않으면 문자열로 처리
- 날짜/시간 형으로 변환을 해주면 일반 문자열에서 사용할 수 없는 추가 메서드를 사용할 수 있음

In [13]:
# pd.read_csv("google_stocks.csv").dtypes
pd.read_csv("google_stocks.csv")

Date      object
Close    float64
dtype: object

 parse_dates 매개변수 : 날짜/시간으로 변환해야하는 텍스트값의 열 이름을 입력하면 변환해줌

In [53]:
pd.read_csv("google_stocks.csv", parse_dates=["Date"]).dtypes
google = pd.read_csv("google_stocks.csv", parse_dates=["Date"], index_col="Date", squeeze=True)
google

<ipython-input-53-4d0fbfd21ac9>:2: FutureWarning: The squeeze argument has been deprecated and will be removed in a future version. Append .squeeze("columns") to the call to squeeze.


  google = pd.read_csv("google_stocks.csv", parse_dates=["Date"], index_col="Date", squeeze=True)


Date
2004-08-19      49.98
2004-08-20      53.95
2004-08-23      54.50
2004-08-24      52.24
2004-08-25      52.80
               ...   
2019-10-21    1246.15
2019-10-22    1242.80
2019-10-23    1259.13
2019-10-24    1260.99
2019-10-25    1265.13
Name: Close, Length: 3824, dtype: float64

독립전쟁 데이터셋 가져오기

- 결측값이 있음
 - NaN 값으로 대체됨

In [32]:
pd.read_csv("revolutionary_war.csv", parse_dates=["Start Date"])

Battle Start Date          State
0                         Powder Alarm 1774-09-01  Massachusetts
1    Storming of Fort William and Mary 1774-12-14  New Hampshire
2     Battles of Lexington and Concord 1775-04-19  Massachusetts
3                      Siege of Boston 1775-04-19  Massachusetts
4                   Gunpowder Incident 1775-04-20       Virginia
..                                 ...        ...            ...
227                Siege of Fort Henry 1782-09-11       Virginia
228         Grand Assault on Gibraltar 1782-09-13            NaN
229          Action of 18 October 1782 1782-10-18            NaN
230          Action of 6 December 1782 1782-12-06            NaN
231          Action of 22 January 1783 1783-01-22       Virginia

[232 rows x 3 columns]

parse_dates 매개변수 :
- mm/dd/yyyy 문자열을 날짜형으로 변환해줌

In [34]:
pd.read_csv("revolutionary_war.csv", parse_dates=["Start Date"], index_col="Start Date")

Battle          State
Start Date                                                  
1774-09-01                       Powder Alarm  Massachusetts
1774-12-14  Storming of Fort William and Mary  New Hampshire
1775-04-19   Battles of Lexington and Concord  Massachusetts
1775-04-19                    Siege of Boston  Massachusetts
1775-04-20                 Gunpowder Incident       Virginia
...                                       ...            ...
1782-09-11                Siege of Fort Henry       Virginia
1782-09-13         Grand Assault on Gibraltar            NaN
1782-10-18          Action of 18 October 1782            NaN
1782-12-06          Action of 6 December 1782            NaN
1783-01-22          Action of 22 January 1783       Virginia

[232 rows x 2 columns]

usecols 매개변수 : 가져올 열목록을 입력

In [54]:
battles = pd.read_csv("revolutionary_war.csv", parse_dates=["Start Date"], index_col="Start Date", usecols=["Start Date", "State"], squeeze=True)
battles

<ipython-input-54-f38f1675f377>:1: FutureWarning: The squeeze argument has been deprecated and will be removed in a future version. Append .squeeze("columns") to the call to squeeze.


  battles = pd.read_csv("revolutionary_war.csv", parse_dates=["Start Date"], index_col="Start Date", usecols=["Start Date", "State"], squeeze=True)


Start Date
1774-09-01    Massachusetts
1774-12-14    New Hampshire
1775-04-19    Massachusetts
1775-04-19    Massachusetts
1775-04-20         Virginia
                  ...      
1782-09-11         Virginia
1782-09-13              NaN
1782-10-18              NaN
1782-12-06              NaN
1783-01-22         Virginia
Name: State, Length: 232, dtype: object

## 3.2 Series 정렬

### 3.2.1 sort_values 메서드를 사용하여 값 기준으로 정렬

구글의 최저 및 최고 주가가 궁금한 경우

sort_values 메서드 : 오름차순으로 정렬된 새 Series 반환

In [58]:
google.sort_values()

Date
2004-09-03      49.82
2004-09-01      49.94
2004-08-19      49.98
2004-09-02      50.57
2004-09-07      50.60
               ...   
2019-04-23    1264.55
2019-10-25    1265.13
2018-07-26    1268.33
2019-04-26    1272.18
2019-04-29    1287.58
Name: Close, Length: 3824, dtype: float64

In [60]:
pokemon.sort_index()

Pokemon
Abomasnow        Grass / Ice
Abra                 Psychic
Absol                   Dark
Accelgor                 Bug
Aegislash      Steel / Ghost
                  ...       
Zoroark                 Dark
Zorua                   Dark
Zubat        Poison / Flying
Zweilous       Dark / Dragon
Zygarde      Dragon / Ground
Name: Type, Length: 809, dtype: object

In [61]:
pd.Series(["AA", "aa", "BB"]).sort_values()

0    AA
2    BB
1    aa
dtype: object

ascending 매개변수 : 정렬 순서를 설정
- 기본이 오름차순(True)
- 내림차순(False)

In [66]:
google.sort_values(ascending=False)
pokemon.sort_index(ascending=False)

Pokemon
Bulbasaur      Grass / Poison
Ivysaur        Grass / Poison
Venusaur       Grass / Poison
Charmander               Fire
Charmeleon               Fire
                    ...      
Stakataka        Rock / Steel
Blacephalon      Fire / Ghost
Zeraora              Electric
Meltan                  Steel
Melmetal                Steel
Name: Type, Length: 809, dtype: object

In [70]:
battles.sort_values()
# battles

Start Date
1781-09-06    Connecticut
1779-07-05    Connecticut
1777-04-27    Connecticut
1777-09-03       Delaware
1777-05-17        Florida
                 ...     
1782-08-08            NaN
1782-08-25            NaN
1782-09-13            NaN
1782-10-18            NaN
1782-12-06            NaN
Name: State, Length: 232, dtype: object

na_position 매개변수 : NaN 값의 배치
- 기본이 "last"
- "first"

In [71]:
battles.sort_values(na_position="first")

Start Date
1775-09-17         NaN
1775-12-31         NaN
1776-03-03         NaN
1776-03-25         NaN
1776-05-18         NaN
                ...   
1781-07-06    Virginia
1781-07-01    Virginia
1781-06-26    Virginia
1781-04-25    Virginia
1783-01-22    Virginia
Name: State, Length: 232, dtype: object

dropna 메서드 : 결측치 제거한 새 Series 반환
- Series의 인덱스가 아닌 Value가 결측치인 것을 제거

In [72]:
battles.dropna().sort_values()

Start Date
1781-09-06    Connecticut
1779-07-05    Connecticut
1777-04-27    Connecticut
1777-09-03       Delaware
1777-05-17        Florida
                 ...     
1781-07-06       Virginia
1781-07-01       Virginia
1781-06-26       Virginia
1781-04-25       Virginia
1783-01-22       Virginia
Name: State, Length: 162, dtype: object

### 3.2.2 sort_index 메서드를 사용해서 인덱스 기준으로 정렬

sort_index 메서드 : 인덱스를 기준으로 Series 정렬

In [75]:
pokemon.sort_index(ascending=True)

Pokemon
Abomasnow        Grass / Ice
Abra                 Psychic
Absol                   Dark
Accelgor                 Bug
Aegislash      Steel / Ghost
                  ...       
Zoroark                 Dark
Zorua                   Dark
Zubat        Poison / Flying
Zweilous       Dark / Dragon
Zygarde      Dragon / Ground
Name: Type, Length: 809, dtype: object

날짜 데이터의 오름차순 정렬은 예전 날짜에서 최근 날짜로
- NaT : not a time

In [81]:
battles.sort_index().dropna()

Start Date
1774-09-01     Massachusetts
1774-12-14     New Hampshire
1775-04-19     Massachusetts
1775-04-19     Massachusetts
1775-04-20          Virginia
                   ...      
1782-08-26    South Carolina
1782-09-11          Virginia
1783-01-22          Virginia
NaT               New Jersey
NaT                 Virginia
Name: State, Length: 162, dtype: object

na_position 매개변수 : 결측치 배치
- 기본 : "last"

In [84]:
battles.sort_index(na_position="first")

Start Date
NaT              New Jersey
NaT                Virginia
NaT                     NaN
NaT                     NaN
1774-09-01    Massachusetts
                  ...      
1782-09-11         Virginia
1782-09-13              NaN
1782-10-18              NaN
1782-12-06              NaN
1783-01-22         Virginia
Name: State, Length: 232, dtype: object

### 3.2.3 nsmallest와 nlargest 메서드를 사용하여 가장 작은 값과 가장 큰 값 검색

구글 주가가 가장 좋았던 날짜 5개를 찾고 싶다면?

풀이법 1)
- 값을 기준으로 내림차순 정렬하여 제일 앞 5개 추출

풀이법 2)
- nlargest 메서드 : Series에서 가장 큰 값을 반환
- n 매개변수 : 반환 갯수 설정(기본 : 5)
- 반환된 Series는 내림차순으로 정렬되어 있음

In [87]:
google.sort_values(ascending=False).head().index

DatetimeIndex(['2019-04-29', '2019-04-26', '2018-07-26', '2019-10-25',
               '2019-04-23'],
              dtype='datetime64[ns]', name='Date', freq=None)

In [91]:
google.nlargest()
google.nlargest(n=5).index

DatetimeIndex(['2019-04-29', '2019-04-26', '2018-07-26', '2019-10-25',
               '2019-04-23'],
              dtype='datetime64[ns]', name='Date', freq=None)

In [93]:
# pokemon.nlargest() # 수치값을 가지는 경우가 아니면 에러
pokemon

Pokemon
Bulbasaur      Grass / Poison
Ivysaur        Grass / Poison
Venusaur       Grass / Poison
Charmander               Fire
Charmeleon               Fire
                    ...      
Stakataka        Rock / Steel
Blacephalon      Fire / Ghost
Zeraora              Electric
Meltan                  Steel
Melmetal                Steel
Name: Type, Length: 809, dtype: object

In [94]:
google.nsmallest(n=5)

Date
2004-09-03    49.82
2004-09-01    49.94
2004-08-19    49.98
2004-09-02    50.57
2004-09-07    50.60
Name: Close, dtype: float64

## inplace 매개변수로 Series 덮어쓰기



In [95]:
battles.head(3)

Start Date
1774-09-01    Massachusetts
1774-12-14    New Hampshire
1775-04-19    Massachusetts
Name: State, dtype: object

In [96]:
battles.sort_values().head(3)

Start Date
1781-09-06    Connecticut
1779-07-05    Connecticut
1777-04-27    Connecticut
Name: State, dtype: object

inplace 매개변수 : Series 객체를 수정할 때 사용
- 기본값 : False
- 수정하고 싶을 때 : True

In [97]:
battles.head(3)

Start Date
1774-09-01    Massachusetts
1774-12-14    New Hampshire
1775-04-19    Massachusetts
Name: State, dtype: object

In [101]:
# battles.sort_values(inplace=True) # 덮어쓰기 됨
battles.head(3)

Start Date
1781-09-06    Connecticut
1779-07-05    Connecticut
1777-04-27    Connecticut
Name: State, dtype: object

주의사항
- 기존 방식이 매번 반환값으로 Series 객체를 생성하여 반환하는 것 때문에 메모리 사용량을 걱정하여 "inplace=True"로 사용하여 반환없이 수정하고 싶어도 아무런 효과가 없음
 - 판다스는 항상 복사본을 생성함

아래의 두 명령어는 동일함

In [ ]:
battles.sort_values(inplace=True)
battles = battles.sort_values()

inplace 매개변수는 사용을 권장하지 않음

## 3.4 value_counts 메서드로 값 계산하기

포켓몬의 가장 흔한 유형은 무엇인가?
- 값들의 빈도수를 카운팅해야함

value_counts 메서드 : Series의 Value들의 발생횟수를 카운팅해서 새 Series로 반환
- 인덱스 레이블 : 기존 Series의 Value
- 값 : 기존 Series의 Value 빈도수

In [102]:
pokemon.value_counts()

Normal                65
Water                 61
Grass                 38
Psychic               35
Fire                  30
                      ..
Fire / Psychic         1
Normal / Ground        1
Psychic / Fighting     1
Dark / Ghost           1
Fire / Ghost           1
Name: Type, Length: 159, dtype: int64

당연히 value_counts 메서드의 반환 Series의 길이는 nunique 메서드의 결과와 같다.

In [104]:
len(pokemon.value_counts())

159

In [107]:
pokemon.nunique()

159

ascending 매개변수 :
- 기본 : False(내림차순)
- 오름차순 : True

In [108]:
pokemon.value_counts(ascending=True)

Fire / Ghost         1
Fighting / Dark      1
Fighting / Steel     1
Normal / Ground      1
Fire / Psychic       1
                    ..
Fire                30
Psychic             35
Grass               38
Water               61
Normal              65
Name: Type, Length: 159, dtype: int64

포켓몬의 각 유형별 비율을 알고 싶을 때
- normalize 매개변수 : True로 설정하면 빈도수를 비율로 반환

In [117]:
(pokemon.value_counts(normalize=True)*100).round(2)

Normal                8.03
Water                 7.54
Grass                 4.70
Psychic               4.33
Fire                  3.71
                      ... 
Fire / Psychic        0.12
Normal / Ground       0.12
Psychic / Fighting    0.12
Dark / Ghost          0.12
Fire / Ghost          0.12
Name: Type, Length: 159, dtype: float64

round() : 소숫점 몇 번째 자리까지 남겨두고 반올림

Google의 주가를 고유값별로 빈도 카운팅

In [119]:
google.value_counts()

287.68     3
194.27     3
307.10     3
288.92     3
290.41     3
          ..
292.43     1
293.41     1
296.75     1
295.46     1
1265.13    1
Name: Close, Length: 3696, dtype: int64

숫자 데이터에 대한 고유값 빈도수 카운팅의 경우, 구간으로 나눠서 빈도수를 구하는 것이 더 유용함
- 데이터의 범위를 확인해야함

In [120]:
google.max()

1287.58

In [121]:
google.min()

49.82

실제 범위는 49.82에서 1287.58 이지만 0~1400으로 잡고, 200단위로 구간을 나눠보자.
- bins 매개변수 : 구간을 리스트의 값으로 설정함
 - 리스트 값의 모든 인접한 쌍을 각 구간의 하한~상한으로 사용
 - 반환 Series는 빈도수에 따라 내림차순으로 정렬됨

In [123]:
buckets = [0, 200, 400, 600, 800, 1000, 1200, 1400]
google.value_counts(bins=buckets)

(200.0, 400.0]      1568
(-0.001, 200.0]      595
(400.0, 600.0]       575
(1000.0, 1200.0]     406
(600.0, 800.0]       380
(800.0, 1000.0]      207
(1200.0, 1400.0]      93
Name: Close, dtype: int64

빈도수가 아닌 인덱스에 따른 정렬을 하려면
- sort_index 메서드

In [124]:
buckets = [0, 200, 400, 600, 800, 1000, 1200, 1400]
google.value_counts(bins=buckets).sort_index()

(-0.001, 200.0]      595
(200.0, 400.0]      1568
(400.0, 600.0]       575
(600.0, 800.0]       380
(800.0, 1000.0]      207
(1000.0, 1200.0]     406
(1200.0, 1400.0]      93
Name: Close, dtype: int64

sort 매개변수 : value_counts의 반환 Series에 대한 정렬 여부
- 기본 : True(값을 기준으로 내림차순 정렬함)
- False : 인덱스를 기준으로 오름차순 정렬

In [125]:
google.value_counts(bins=buckets, sort=False)

(-0.001, 200.0]      595
(200.0, 400.0]      1568
(400.0, 600.0]       575
(600.0, 800.0]       380
(800.0, 1000.0]      207
(1000.0, 1200.0]     406
(1200.0, 1400.0]      93
Name: Close, dtype: int64

(-0.001, 200.0] : -0.001보다 크고, 200 이하라는 의미
- 소괄호('(')는 개구간으로 값은 제외함을 의미
- 대괄호(']')는 폐구간으로 값을 포함함을 의미
 - "(-0.001, 200.0]"는 "반개구간"임

bins 매개변수 : 정수도 입력가능
- 최댓값과 최솟값의 차이를 자동으로 계산해서, 입력받은 정수의 갯수만큼 등간격으로 구간들을 만들어줌

In [126]:
google.value_counts(bins=6, sort=False)

(48.581, 256.113]      1204
(256.113, 462.407]     1104
(462.407, 668.7]        507
(668.7, 874.993]        380
(874.993, 1081.287]     292
(1081.287, 1287.58]     337
Name: Close, dtype: int64

battles 데이터셋을 다뤄보자.

In [127]:
battles

Start Date
1781-09-06    Connecticut
1779-07-05    Connecticut
1777-04-27    Connecticut
1777-09-03       Delaware
1777-05-17        Florida
                 ...     
1782-08-08            NaN
1782-08-25            NaN
1782-09-13            NaN
1782-10-18            NaN
1782-12-06            NaN
Name: State, Length: 232, dtype: object

독립전쟁에서 가장 많은 전투를 벌인 주는?

In [129]:
battles.value_counts()

South Carolina    31
New York          28
New Jersey        24
Virginia          21
Massachusetts     11
Pennsylvania      10
North Carolina     9
Florida            8
Georgia            6
Rhode Island       3
Vermont            3
Connecticut        3
Indiana            1
Louisiana          1
New Hampshire      1
Delaware           1
Ohio               1
Name: State, dtype: int64

dropna = False : NaN값을 포함해서 카운팅함

In [130]:
battles.value_counts(dropna=False)

NaN               70
South Carolina    31
New York          28
New Jersey        24
Virginia          21
Massachusetts     11
Pennsylvania      10
North Carolina     9
Florida            8
Georgia            6
Rhode Island       3
Vermont            3
Connecticut        3
Indiana            1
Louisiana          1
New Hampshire      1
Delaware           1
Ohio               1
Name: State, dtype: int64

독립전쟁 중 가장 많은 전투가 일어난 날짜는?

In [133]:
battles.index.value_counts()

1781-04-25    2
1781-04-15    2
1777-08-22    2
1782-01-11    2
1775-04-19    2
             ..
1776-09-16    1
1781-10-25    1
1777-07-08    1
1779-07-22    1
1782-12-06    1
Name: Start Date, Length: 217, dtype: int64

## 3.5 apply 메서드를 사용하여 모든 Series 값에 대한 함수 호출

In [134]:
round(99.5)

100

In [138]:
google.apply(round)

Date
2004-08-19      50
2004-08-20      54
2004-08-23      54
2004-08-24      52
2004-08-25      53
              ... 
2019-10-21    1246
2019-10-22    1243
2019-10-23    1259
2019-10-24    1261
2019-10-25    1265
Name: Close, Length: 3824, dtype: int64

한 가지 유형을 가지는 포켓몬과 두 가지 유형을 가지는 포켓몬이 각각 몇마리인지 알고 싶은 경우

In [139]:
def single_or_multi(ptype):
  if "/" in ptype:
    return "Multi"
  return "Single"

In [141]:
pokemon.apply(single_or_multi).value_counts()

Multi     405
Single    404
Name: Type, dtype: int64

## 3.6 코딩 챌린지

In [143]:
import datetime as dt
today = dt.datetime(2023, 10, 13)
today.strftime("%A")

'Friday'

In [144]:
def day_of_week(date):
  return date.strftime("%A")

In [155]:
days_of_war = pd.read_csv("revolutionary_war.csv", parse_dates=["Start Date"], usecols=["Start Date"], squeeze=True)
days_of_war

<ipython-input-155-62776a7cc91b>:1: FutureWarning: The squeeze argument has been deprecated and will be removed in a future version. Append .squeeze("columns") to the call to squeeze.


  days_of_war = pd.read_csv("revolutionary_war.csv", parse_dates=["Start Date"], usecols=["Start Date"], squeeze=True)


0     1774-09-01
1     1774-12-14
2     1775-04-19
3     1775-04-19
4     1775-04-20
         ...    
227   1782-09-11
228   1782-09-13
229   1782-10-18
230   1782-12-06
231   1783-01-22
Name: Start Date, Length: 232, dtype: datetime64[ns]

In [158]:
days_of_war.dropna().apply(day_of_week).value_counts()

Saturday     39
Friday       39
Wednesday    32
Thursday     31
Sunday       31
Tuesday      29
Monday       27
Name: Start Date, dtype: int64